### PR TITLE
Instance initialization improved

### PR DESCRIPTION
**Pull request by Heromyth**

### CODE REVIEW

File: source/poodinis/polyfill.d [hunk 1]
```diff
@@ -43,21 +43,24 @@ static if (!__traits(compiles, basicExceptionCtors)) {
...
         }
         else static if (is(X[0] T))
         {
             // x is a type.  Take the type of it and examine.
-            enum isFunction = is(T == function);
+            enum isFunction1 = is(T == function);
         }
         else
-            enum isFunction = false;
+            enum isFunction1 = false;
```
Comment (mbierlee) [inline source/poodinis/polyfill.d]: Why did you rename this trait? There is no std.trait isFunction1. So now you're always polyfilling even when isFunction is supported by the current dmd version

File: example/constructorinjection/app.d [hunk 2]
```diff
@@ -57,7 +64,10 @@ void main() {
 	auto dependencies = new shared DependencyContainer();
 	dependencies.register!Scheduler;
 	dependencies.register!Calendar;
-	dependencies.register!HardwareClock;
+	dependencies.register!HardwareClock( {
+		writeln("Running the creator");
+		return new HardwareClock("clock name");
+	});
```
Comment (mbierlee) [inline example/constructorinjection/app.d]: Perhaps create a separate example for this. This example is focused on injection via constructors, not creation.

File: example/postconstructorpredestructor/app.d [hunk 3]
```diff
@@ -42,8 +42,14 @@ class AClass {
...
-	container.register!ADependency;
-	container.register!AClass;
+	container.register!(ADependency).onConstructed((Object obj) {
+		writeln("ADependency constructed");
+	});
+
+	container.register!(AClass).onConstructed((Object obj) {
+		writeln("AClass constructed");
+	});
+
```
Comment (mbierlee) [inline example/postconstructorpredestructor/app.d]: Could also be put in its own example. Examples try to showcase one thing only to teach people topic-by-topic.

File: source/poodinis/container.d [hunk 4]
```diff
@@ -192,6 +201,44 @@ synchronized class DependencyContainer {
...
+        static if (!is(SuperType == ConcreteType)) {
+            if (!hasOption(options, persistentRegistrationOptions, RegistrationOption.doNotAddConcreteTypeRegistration)) {
+                auto concreteTypeRegistration = register!ConcreteType;
+                concreteTypeRegistration.linkTo(newRegistration);
+            }
+        }
+
+        registrations[registeredType] ~= cast(shared(Registration)) newRegistration;
+        return newRegistration;
+    }
```
Comment (mbierlee) [inline source/poodinis/container.d]: This function is now duplicating a lot of code from
public Registration register(SuperType, ConcreteType : SuperType)(
            RegistrationOption options = RegistrationOption.none) if (!is(ConcreteType == struct)) 

Is there a way to call this function in that one?

File: source/poodinis/container.d [hunk 5]
```diff
@@ -192,6 +201,44 @@ synchronized class DependencyContainer {
...
+        }
+
+        InstanceFactory instanceFactory = new class InstanceFactory {
+            protected override Object createInstance() { 
+                return creator();
+            }
+        };
+
+        auto newRegistration = new AutowiredRegistration!ConcreteType(registeredType, instanceFactory, this);
+        newRegistration.singleInstance();
```
Comment (mbierlee) [inline source/poodinis/container.d]: This doesn't make much sense anymore, the custom instance factory doesn't guarentee a singleInstance anymore.
Instead of supplying the factory in the signature of register(), is is possible to supply the factory like singleInstance(), newInstance() etc does?

Example:
`dependencies.register!Driver.createdBy( {
	return new Driver();
})`

Not particularly married to the "createdBy" name

File: source/poodinis/container.d [hunk 6]
```diff
@@ -192,6 +201,44 @@ synchronized class DependencyContainer {
         return newRegistration;
     }
 
+    /**
+     * 
+     */
```
Comment (mbierlee) [inline source/poodinis/container.d]: Could you finish the ddoc?

File: source/poodinis/container.d [hunk 7]
```diff
@@ -326,6 +373,12 @@ synchronized class DependencyContainer {
         }
     }
 
+    bool isRegistered(RegistrationType)() {
```
Comment (mbierlee) [inline source/poodinis/container.d]: Could you add some ddoc here?

**Comment by mbierlee** [comment]
Love the addition of a custom initializer!

If you don't get to the docs, tutorial and example changes don't worry. Let me know and I will take care of that.
My main blocker for merging right now is the interface of the InjectionInitializer and the function1 refactoring.

**Comment by Heromyth** [comment]
Sorry for the delay. We are busy with the improvements of Hunt-Framework which depends on poodinis [heavily](https://github.com/huntlabs/hunt-framework/tree/master/source/hunt/framework/provider). Can you merge this at first and release a beta version. Thanks.

**Comment by mbierlee** [comment]
> Sorry for the delay. We are busy with the improvements of Hunt-Framework which depends on poodinis [heavily](https://github.com/huntlabs/hunt-framework/tree/master/source/hunt/framework/provider). Can you merge this at first and release a beta version. Thanks.

Sure thing! Expect version 8.1.0-beta to be available soon

**Comment by mbierlee** [comment]
Release with v8.1.0-beta
https://github.com/mbierlee/poodinis/releases/tag/v8.1.0-beta

**Comment by mbierlee** [comment]
By the way you could've also used application contexts, see https://github.com/mbierlee/poodinis/blob/master/example/applicationcontext/app.d
usage of UDA @Component

I'll keep this in for those who don't want to set-up the ApplicationContext boilerplate

**Comment by Heromyth** [comment]
> 
> 
> By the way you could've also used application contexts, see https://github.com/mbierlee/poodinis/blob/master/example/applicationcontext/app.d
> usage of UDA `@Component`
> 
> I'll keep this in for those who don't want to set-up the ApplicationContext boilerplate

Yes, the ApplicationContext does somthing alike. We just follow the rule of laravel, see https://laravel.com/docs/5.6/providers.

**Comment by Heromyth** [comment]
> 
> 
> Released with v8.1.0-beta
> https://github.com/mbierlee/poodinis/releases/tag/v8.1.0-beta

Thanks.
